pkg/webhook/external/generic_list_options.go

```diff
@@ -0,0 +1,151 @@
+package external
+
+import (
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+	"github.com/akuity/kargo/pkg/expressions"
+	"github.com/akuity/kargo/pkg/expressions/function"
+)
+
+// buildListOptionsForTarget builds a list of client.ListOption based on the
+// provided GenericWebhookTarget's selectors. The returned ListOptions can be
+// used to list Kubernetes resources that match the target's selection criteria.
+func buildListOptionsForTarget(
+	project string,
+	t kargoapi.GenericWebhookTarget,
+	env map[string]any,
+) ([]client.ListOption, error) {
+	listOpts := []client.ListOption{client.InNamespace(project)}
+	if len(t.IndexSelector.MatchExpressions) > 0 {
+		indexSelectorListOpts, err := newListOptionsForIndexSelector(t.IndexSelector, env)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create field selector: %w", err)
+		}
+		listOpts = append(listOpts, indexSelectorListOpts...)
+	}
+	if len(t.LabelSelector.MatchLabels) > 0 || len(t.LabelSelector.MatchExpressions) > 0 {
+		labelSelectorListOpts, err := newListOptionsForLabelSelector(t.LabelSelector, env)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create label selector: %w", err)
+		}
+		listOpts = append(listOpts, labelSelectorListOpts...)
+	}
+	if len(listOpts) == 1 {
+		listOpts = append(listOpts, client.MatchingLabelsSelector{
+			Selector: labels.Everything(),
+		})
+	}
+	return listOpts, nil
+}
+
+// newListOptionsForIndexSelector creates a list of client.ListOption based on
+// the provided IndexSelector and environment for expression evaluation.
+func newListOptionsForIndexSelector(
+	is kargoapi.IndexSelector,
+	env map[string]any,
+) ([]client.ListOption, error) {
+	var listOpts []client.ListOption
+	for _, expr := range is.MatchExpressions {
+		resultStr, err := evalAsString(expr.Value, env)
+		if err != nil {
+			return nil, fmt.Errorf("failed to evaluate values expression as string: %w", err)
+		}
+		var s fields.Selector
+		switch expr.Operator {
+		case kargoapi.IndexSelectorRequirementOperatorEqual:
+			s = fields.OneTermEqualSelector(expr.Key, resultStr)
+		case kargoapi.IndexSelectorRequirementOperatorNotEqual:
+			s = fields.OneTermNotEqualSelector(expr.Key, resultStr)
+		default:
+			return nil, fmt.Errorf("unsupported operator %q in index selector expression", expr.Operator)
+		}
+		listOpts = append(listOpts, client.MatchingFieldsSelector{Selector: s})
+	}
+	return listOpts, nil
+}
+
+// newListOptionsForLabelSelector creates a list of client.ListOption based on
+// the provided LabelSelector.
+func newListOptionsForLabelSelector(ls metav1.LabelSelector, env map[string]any) ([]client.ListOption, error) {
+	var requirements []labels.Requirement
+	for _, e := range ls.MatchExpressions {
```

Comment:
Applies in several spots. More descriptive variable names would make all this code a lot easier to comprehend.